videoinput.c

```diff
@@ -564,9 +635,23 @@ inline void cVideoInput::Receive(const uchar *data, int length)
   m_VideoBuffer->Put(data, length);
 }
 
-void cVideoInput::Retune()
+void cVideoInput::RequestRetune()
+{
+  m_RetuneRequested = true;
+  m_Event.Signal();
+}
+
+cVideoInput::eReceivingStatus cVideoInput::ReceivingStatus()
 {
-  cMutexLock lock(&m_Mutex);
-  m_IsRetune = true;
-  m_Event.Broadcast();
+  if (!m_Device || !m_Receiver || !m_DummyReceiver)
+    return RETUNE;
+  if (m_RetuneRequested || !m_Receiver->IsAttached())
+  {
+    (void)m_Device->Receiving();  // wait for the receivers mutex
+    if (!m_DummyReceiver->IsAttached())  // DetachAllReceivers() was called
+      return CLOSE;
+    else if (!m_PmtChange)
```

Comment:
Receivers operations (`AttachReceiver()`, `DetachXXX()`, `Receiving()`) are protected by a mutex, but `IsAttached()` is not. It just reads `device` value. At the moment we get `m_Receiver->device == NULL` (`!IsAttached()`), `m_DummyReceiver->device` might be not NULL - `DetachAllReceivers()` hasn't yet detached it. That's why we have to wait for `DetachAllReceivers()` to complete. At this point we can be sure that if both receivers has been detached, we get two NULL `device`s. In other words, without `Receiving()` (or any other call that holds the receiver mutex) it's possible to get a false RETUNE when it should have been CLOSE. This would clash with recording.
But `device` members are not atomic/protected as well as `m_RetuneRequested`. We can use our own 'protected' flags set by receiver callbacks instead. I'll rework this and set a fix as a new PR. Do you agree?
Do we still need m_PmtChange?